### PR TITLE
794 kendo template excel export

### PR DESCRIPTION
**Pull request by Crevil**
Implement excel template attribute to kendo grid.
Add domain entity interfaces for used properties.

### CODE REVIEW

File: Presentation.Web/app/components/it-contract/it-contract-overview.controller.ts [hunk 1]
```diff
@@ -302,29 +316,41 @@
...
+                        },
                         attributes: { "class": "text-center" },
                         sortable: false,
-                        filterable: false,
+                        filterable: false
                     },
                     {
                         field: "ItContractId", title: "KontraktID", width: 150,
-                        persistId: "contractid", // DON'T YOU DARE RENAME!
+                        persistId: "contractid", // DON'T YOU DARE RENAME!,
```
Comment (snebjorn) [inline Presentation.Web/app/components/it-contract/it-contract-overview.controller.ts]: Remove trailing ,

File: Presentation.Web/app/components/it-contract/it-contract-overview.controller.ts [hunk 2]
```diff
@@ -360,8 +387,17 @@
                     {
                         field: "AssociatedSystemUsages", title: "IT System", width: 150,
                         persistId: "itsys", // DON'T YOU DARE RENAME!
-                        template: "#: AssociatedSystemUsages.length > 0 ? _.first(AssociatedSystemUsages).ItSystemUsage.ItSystem.Name : '' #" +
-                                    "#= AssociatedSystemUsages.length > 1 ? ' (' + AssociatedSystemUsages.length + ')' : '' #",
+                        template: (dataItem: IItContractOverview) => {
```
Comment (snebjorn) [inline Presentation.Web/app/components/it-contract/it-contract-overview.controller.ts]: Isn't IItContractOverview inferred

File: Presentation.Web/app/components/it-contract/it-contract-overview.controller.ts [hunk 3]
```diff
@@ -386,59 +422,71 @@
                     {
                         field: "Esdh", title: "ESDH ref", width: 150,
                         persistId: "esdh", // DON'T YOU DARE RENAME!
-                        template: "#= Esdh ? '<a target=\"_blank\" href=\"' + Esdh + '\"><i class=\"fa fa-link\"></a>' : '' #",
+                        template: dataItem => dataItem.Esdh ? `<a target=\"_blank\" href=\"${dataItem.Esdh}\"><i class=\"fa fa-link\"></a>` : "",
```
Comment (snebjorn) [inline Presentation.Web/app/components/it-contract/it-contract-overview.controller.ts]: Is it necessary to escape quotes?

File: Presentation.Web/app/components/it-contract/it-contract-overview.controller.ts [hunk 4]
```diff
@@ -386,59 +422,71 @@
...
                                 showOperators: false,
-                                operator: "contains",
+                                operator: "contains"
                             }
                         }
                     },
                     {
                         field: "Acquisition", title: "Anskaffelse", width: 90,
-                        persistId: "acquisition", // DON'T YOU DARE RENAME!
+                        persistId: "acquisition", // DON'T YOU DARE RENAME!,
```
Comment (snebjorn) [inline Presentation.Web/app/components/it-contract/it-contract-overview.controller.ts]: ,

File: Presentation.Web/app/components/it-contract/it-contract-overview.controller.ts [hunk 5]
```diff
@@ -450,42 +498,57 @@
...
+                        filterable: false
                     },
                     {
-                        field: "", title: "Audit status", width: 90,
+                        field: "AuditStatus", title: "Audit status", width: 90,
                         persistId: "auditstatus", // DON'T YOU DARE RENAME!
-                        template: this.audioStatusTemplate,
+                        template: dataItem => {
+                            if (dataItem.status.max > 0) {
+                                var str = JSON.stringify(dataItem.status);
```
Comment (snebjorn) [inline Presentation.Web/app/components/it-contract/it-contract-overview.controller.ts]: Now that I think about it. Then it should probably be $window.JSON

File: Presentation.Web/app/components/it-contract/it-contract-overview.controller.ts [hunk 6]
```diff
@@ -630,11 +679,47 @@
...
                 });
 
+                // render templates
+                var sheet = e.workbook.sheets[0];
+
+                // skip header row
+                for (var rowIndex = 1; rowIndex < sheet.rows.length; rowIndex++) {
+                    var row = sheet.rows[rowIndex];
+
+                    // -1 as sheet has header and dataSource hasn't
```
Comment (snebjorn) [inline Presentation.Web/app/components/it-contract/it-contract-overview.controller.ts]: hasn't -> doesn't

File: Presentation.Web/app/components/it-contract/it-contract-plan.controller.ts [hunk 7]
```diff
@@ -374,7 +401,8 @@
                     {
                         field: "Esdh", title: "ESDH ref", width: 150,
                         persistId: "esdh", // DON'T YOU DARE RENAME!
-                        template: "#= Esdh ? '<a target=\"_blank\" href=\"' + Esdh + '\"><i class=\"fa fa-link\"></a>' : '' #",
+                        template: dataItem => dataItem.Esdh ? `<a target=\"_blank\" href=\"${dataItem.Esdh}\"><i class=\"fa fa-link\"></a>` : "",
```
Comment (snebjorn) [inline Presentation.Web/app/components/it-contract/it-contract-plan.controller.ts]: escaping "?

File: Presentation.Web/app/components/it-contract/it-contract-plan.controller.ts [hunk 8]
```diff
@@ -627,42 +724,61 @@
...
 
-        private roleTemplate(dataItem, roleId) {
-            var roles = "";
+                // skip header row
+                for (var rowIndex = 1; rowIndex < sheet.rows.length; rowIndex++) {
+                    var row = sheet.rows[rowIndex];
 
-            if (dataItem.roles[roleId] === undefined)
-                return roles;
+                    // -1 as sheet has header and dataSource hasn't
```
Comment (snebjorn) [inline Presentation.Web/app/components/it-contract/it-contract-plan.controller.ts]: doesn't

File: Presentation.Web/app/components/it-project/it-project-overview.controller.ts [hunk 9]
```diff
@@ -507,24 +567,25 @@
...
-                        "</label>" +
-                        "</div>" +
-                        '<select data-ng-model="dataItem.PriorityPf" data-autosave="api/itproject/{{dataItem.Id}}" data-field="priorityPf">' +
-                        '<option value="None">-- Vælg --</option>' +
-                        '<option value="High">Høj</option>' +
-                        '<option value="Mid">Mellem</option>' +
-                        '<option value="Low">Lav</option>' +
-                        "</select>",
+                        template: () => `<div class="btn-group btn-group-sm" data-toggle="buttons">
+                                                    <label class="btn btn-star" data-ng-class="{ \'unstarred\': !dataItem.IsPriorityLocked }" data-ng-disabled="!dataItem.hasWriteAccess" data-ng-click="dataItem.IsPriorityLocked = !dataItem.IsPriorityLocked">
```
Comment (snebjorn) [inline Presentation.Web/app/components/it-project/it-project-overview.controller.ts]: escaping of '

File: Presentation.Web/app/components/it-system/it-interface/it-interface-catalog.controller.ts [hunk 10]
```diff
@@ -277,95 +290,120 @@
...
-                                operator: "contains",
+                                operator: "contains"
                             }
-                        },
+                        }
                     },
                     {
                         field: "Organization.Name", title: "Oprettet af: Organisation", width: 150,
-                        persistId: "orgname", // DON'T YOU DARE RENAME!
+                        persistId: "orgname", // DON'T YOU DARE RENAME!,
```
Comment (snebjorn) [inline Presentation.Web/app/components/it-system/it-interface/it-interface-catalog.controller.ts]: ,

File: Presentation.Web/app/kitos.ts [hunk 11]
```diff
@@ -3,22 +3,29 @@
...
+    export interface IKendoGrid<TDataSource> extends kendo.ui.Grid {
+        getOptions(): IKendoGridOptions<TDataSource>;
+        columns: IKendoGridColumn<TDataSource>[];
+    }
+
+    export interface IKendoDataObservableObject extends kendo.data.ObservableObject {
+        id: number;
     }
 
     export interface ILodashWithMixins extends _.LoDashStatic {
```
Comment (snebjorn) [inline Presentation.Web/app/kitos.ts]: Oops. ILodashWithMixins should ILoDashWithMixins

**Comment by snebjorn** [comment]
Good job!

Is it possible to make generic interfaces op-in? So we don't have to do IKendoGrid<any> for example.

**Comment by Crevil** [comment]
As of [typescript issue #2175](https://github.com/Microsoft/TypeScript/issues/2175) generics with default types / optionals are still in discussion.

**Pull request merged**